File: app/src/test/java/no/nav/tilbakemeldingsmottak/ReGenereringAvPdf.java
package no.nav.tilbakemeldingsmottak;

import no.nav.tilbakemeldingsmottak.model.Innmelder;
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklageRequest;
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklageRequest.GjelderSosialhjelpEnum;
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklageRequest.KlagetyperEnum;
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklageRequest.PaaVegneAvEnum;
import no.nav.tilbakemeldingsmottak.model.PaaVegneAvPerson;
import no.nav.tilbakemeldingsmottak.rest.common.pdf.PdfService;
import org.junit.jupiter.api.Test;

import java.io.FileOutputStream;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class ReGenereringAvPdf {


    private OpprettServiceklageRequest createOpprettServiceklageRequestPrivatperson(List<String> klage) {
        return OpprettServiceklageRequest.builder()
                .paaVegneAv(PaaVegneAvEnum.PRIVATPERSON)
                .innmelder(Innmelder.builder()
                        .navn(klage.get(32))
                        .telefonnummer(klage.get(33))
                        .personnummer(klage.get(3))
                        .build())
                .klagetyper(konverterTilKlageType(Arrays.stream(klage.get(4).split(",")).toList()))
                .gjelderSosialhjelp(findSosialhjelpTypeText(klage.get(5)))
                .klagetekst(klage.get(6))
                .oenskerAaKontaktes(klage.get(25) != null)
                .enhetsnummerPaaklaget(klage.get(13))
                .klagetypeUtdypning(klage.get(30))
                .build();
    }

    private LocalDateTime mapTilLocalDateTime(String fremmetText) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy HH.mm.ss");
        return LocalDateTime.parse(fremmetText.substring(0, 19), formatter);
    }

    private List<KlagetyperEnum> konverterTilKlageType(List<String> klageTypeStrenger) {
        List<KlagetyperEnum> typer = new LinkedList<>();
        for (String klageStreng : klageTypeStrenger) {
            typer.add(findByText(klageStreng.trim()));
        }
        return typer;
    }

    private KlagetyperEnum findByText(String text) {
        for (KlagetyperEnum v : KlagetyperEnum.values()) {
            if (v.value.equalsIgnoreCase(text)) {
                return v;
            }
        }
        return null;
    }

    private GjelderSosialhjelpEnum findSosialhjelpTypeText(String text) {
        for (GjelderSosialhjelpEnum t : GjelderSosialhjelpEnum.values()) {
            if (t.value.equalsIgnoreCase(text)) {
                return t;
            }
        }
        return GjelderSosialhjelpEnum.VET_IKKE;
    }

    private OpprettServiceklageRequest createOpprettServiceklageRequestPaaVegneAvPerson(List<String> klage) {
        return OpprettServiceklageRequest.builder()
                .paaVegneAv(PaaVegneAvEnum.ANNEN_PERSON)
                .innmelder(Innmelder.builder()
                        .navn(klage.get(32))
                        .telefonnummer(klage.get(33))
                        .harFullmakt(true)
                        .personnummer(klage.get(3))
                        .rolle("Fullmektig")
                        .build())
                .paaVegneAvPerson(PaaVegneAvPerson.builder()
                        .navn(klage.get(34))
                        .personnummer(klage.get(35))
                        .build())
                .klagetyper(konverterTilKlageType(Arrays.stream(klage.get(4).split(",")).toList()))
                .klagetekst(klage.get(6))
                .oenskerAaKontaktes(klage.get(25) != null)
                .build();
    }

    @Test
    public void generatePdfs() {
        try {
            Map<Integer, List<String>> serviceklager = new ReadXslxFile().readExcelFile("serviceklage-eks.xlsx");
            PdfService pdfService = new PdfService();

            boolean header = true;
            for (List<String> klage : serviceklager.values()) {
                if (header) {
                    header = false;
                    continue;
                }
                if ("Bruker selv som privatperson".equalsIgnoreCase(klage.get(10))) {
                    OpprettServiceklageRequest request = createOpprettServiceklageRequestPrivatperson(klage);
                    byte[] klagePdf = pdfService.opprettServiceklagePdf(request, "1".equals(klage.get(31)), mapTilLocalDateTime(klage.get(2)));
                    //writeBytesToFile(klagePdf, "src/test/resources/serviceklage-jp-"+ klage.get(1) +".pdf");
                    assertTrue(klagePdf != null && klagePdf.length > 0);
                } else if ("På vegne av en annen privatperson".equalsIgnoreCase(klage.get(10))) {
                    OpprettServiceklageRequest request = createOpprettServiceklageRequestPaaVegneAvPerson(klage);
                    byte[] klagePdf = pdfService.opprettServiceklagePdf(request, "1".equals(klage.get(31)), mapTilLocalDateTime(klage.get(2)));
                    //writeBytesToFile(klagePdf, "src/test/resources/serviceklage-jp-"+ klage.get(1) +".pdf");
                    assertTrue(klagePdf != null && klagePdf.length > 0);
                }
            }

        } catch (Exception e) {
            throw new RuntimeException("Lesing fra excel fil feilet" + e.getMessage());
        }
    }


    private void writeBytesToFile(byte[] data, String filePath) throws IOException {
        try (FileOutputStream fos = new FileOutputStream(filePath)) {
            fos.write(data);
        }
    }


}

File: app/src/test/kotlin/no/nav/tilbakemeldingsmottak/util/ReGenereringAvPdf.kt
package no.nav.tilbakemeldingsmottak.util

import no.nav.tilbakemeldingsmottak.model.*
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklageGjelderSosialhjelp.VET_IKKE
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklagePaaVegneAv.ANNEN_PERSON
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklagePaaVegneAv.PRIVATPERSON
import no.nav.tilbakemeldingsmottak.rest.common.pdf.PdfService
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.io.FileOutputStream
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*

@kotlin.ExperimentalStdlibApi
internal class ReGenereringAvPdf {
    private fun createOpprettServiceklageRequestPrivatperson(klage: List<String?>): OpprettServiceklageRequest {
        return OpprettServiceklageRequest(
            paaVegneAv = PRIVATPERSON,
            innmelder = Innmelder(navn = klage[32], telefonnummer = klage[33], personnummer = klage[3]),
            klagetyper = konverterTilKlageType(
                Arrays.stream(
                    klage[4]!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }
                        .toTypedArray<String>()).toList()),
            gjelderSosialhjelp = findSosialhjelpTypeText(klage[5]),
            klagetekst = klage[6],
            oenskerAaKontaktes = klage[25] != null,
            enhetsnummerPaaklaget = klage[13],
            klagetypeUtdypning = klage[30]
        )
    }

    private fun mapTilLocalDateTime(fremmetText: String?): LocalDateTime {
        val formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy HH.mm.ss")
        return LocalDateTime.parse(fremmetText!!.substring(0, 19), formatter)
    }

    private fun konverterTilKlageType(klageTypeStrenger: List<String>): List<OpprettServiceklageKlagetype> {
        val typer: MutableList<OpprettServiceklageKlagetype?> = LinkedList()
        for (klageStreng in klageTypeStrenger) {
            typer.add(findByText(klageStreng.trim { it <= ' ' }))
        }
        return typer.filterNotNull()
    }

    private fun findByText(text: String): OpprettServiceklageKlagetype? {
        for (v in OpprettServiceklageKlagetype.entries) {
            if (v.value.equals(text, ignoreCase = true)) {
                return v
            }
        }
        return null
    }

    private fun findSosialhjelpTypeText(text: String?): OpprettServiceklageGjelderSosialhjelp {
        for (t in OpprettServiceklageGjelderSosialhjelp.entries) {
            if (t.value.equals(text, ignoreCase = true)) {
                return t
            }
        }
        return VET_IKKE
    }

    private fun createOpprettServiceklageRequestPaaVegneAvPerson(klage: List<String?>): OpprettServiceklageRequest {
        return OpprettServiceklageRequest(paaVegneAv = ANNEN_PERSON,
            innmelder = Innmelder(
                navn = klage[32],
                telefonnummer = klage[33],
                harFullmakt = true,
                personnummer = klage[3],
                rolle = "Fullmektig"
            ),
            paaVegneAvPerson = PaaVegneAvPerson(navn = klage[34], personnummer = klage[35]),
            klagetyper = konverterTilKlageType(
                Arrays.stream(
                    klage[4]!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }
                        .toTypedArray<String>()).toList()
            ),
            klagetekst = klage[6],
            oenskerAaKontaktes = klage[25] != null)
    }

    @Test
    fun generatePdfs() {
        try {
            val serviceklager = ReadXslxFile().readExcelFile("serviceklage-eks.xlsx")
            val pdfService = PdfService()
            var header = true
            for (klage in serviceklager.values) {
                if (header) {
                    header = false
                    continue
                }
                if ("Bruker selv som privatperson".equals(klage[10], ignoreCase = true)) {
                    val request = createOpprettServiceklageRequestPrivatperson(klage)
                    val klagePdf = pdfService.opprettServiceklagePdf(
                        request, "1" == klage[31], mapTilLocalDateTime(
                            klage[2]
                        )
                    )
                    //writeBytesToFile(klagePdf, "src/test/resources/serviceklage-jp-"+ klage.get(1) +".pdf");
                    assertTrue(klagePdf.isNotEmpty())
                } else if ("På vegne av en annen privatperson".equals(klage[10], ignoreCase = true)) {
                    val request = createOpprettServiceklageRequestPaaVegneAvPerson(klage)
                    val klagePdf = pdfService.opprettServiceklagePdf(
                        request, "1" == klage[31], mapTilLocalDateTime(
                            klage[2]
                        )
                    )
                    //writeBytesToFile(klagePdf, "src/test/resources/serviceklage-jp-"+ klage.get(1) +".pdf");
                    assertTrue(klagePdf.isNotEmpty())
                }
            }
        } catch (e: Exception) {
            throw RuntimeException("Lesing fra excel fil feilet" + e.message)
        }
    }

    private fun writeBytesToFile(data: ByteArray, filePath: String) {
        FileOutputStream(filePath).use { fos -> fos.write(data) }
    }
}
